File: stockflow-frontend/src/features/purchase/services/purchaseService.js
import api from '@/services/api';
import { API } from '@/constants/api';

/**
 * purchaseService — API integration service for Purchase Management.
 */
export const purchaseService = {
  getPaginated: async (page = 0, size = 10, sortBy = 'purchaseDate', direction = 'DESC') => {
    const res = await api.get(`${API.PURCHASES}/page`, {
      params: { page, size, sort: sortBy, direction },
    });
    return res.data;
  },

  getAllList: async () => {
    const res = await api.get(API.PURCHASES);
    return res.data;
  },

  getById: async (id) => {
    const res = await api.get(`${API.PURCHASES}/${id}`);
    return res.data;
  },

  search: async (params = {}, page = 0, size = 10, sortBy = 'purchaseDate', direction = 'DESC') => {
    const res = await api.get(`${API.PURCHASES}/search`, {
      params: {
        purchaseNumber: params.purchaseNumber || undefined,
        supplierName: params.supplierName || undefined,
        invoiceNumber: params.invoiceNumber || undefined,
        productKeyword: params.productKeyword || undefined,
        page,
        size,
        sort: sortBy,
        direction,
      },
    });
    return res.data;
  },

  create: async (data) => {
    const res = await api.post(API.PURCHASES, data);
    return res.data;
  },
};

File: stockflow-backend/src/main/java/com/stockflow/backend/service/PurchaseService.java
package com.stockflow.backend.service;

import com.stockflow.backend.dto.purchase.PurchaseRequest;
import com.stockflow.backend.dto.purchase.PurchaseResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;

/**
 * PurchaseService — Business operations interface for Purchase module.
 */
public interface PurchaseService {

    /** Create a new purchase order and automatically update inventory (STOCK_IN) */
    PurchaseResponse createPurchase(PurchaseRequest request, String username);

    /** Get purchase by ID */
    PurchaseResponse getPurchaseById(Long id);

    /** Get all purchases (flat list) */
    List<PurchaseResponse> getAllPurchases();

    /** Get paginated purchases */
    Page<PurchaseResponse> getAllPurchasesPaginated(Pageable pageable);

    /** Search purchases */
    Page<PurchaseResponse> searchPurchases(
            String purchaseNumber,
            String supplierName,
            String invoiceNumber,
            String productKeyword,
            LocalDateTime dateFrom,
            LocalDateTime dateTo,
            Pageable pageable);

    /** Filter purchases */
    Page<PurchaseResponse> filterPurchases(
            Long supplierId,
            Long warehouseId,
            String paymentStatus,
            LocalDateTime dateFrom,
            LocalDateTime dateTo,
            Pageable pageable);
}
